### ISSUE TITLE
Supervised Multivariate HMM learning using IndependentComponentDistribution 

### ISSUE THREAD

**Issue by mmrahimi** [jmschrei/pomegranate#944]
Hi @jmschrei, 

Thank you for your wonderful work on this library.

I am trying to set up supervised learning on a couple of existing sequences with multivariate data. 

My input's shape is as follows: 10 (number of days -- i.e., number of sequences) x 24 (hours) x 3 (features)

This is my current approach to generate the model:

```
hidden_states_types = ['a','b','c','d','nan']
labels = [['model-start']+x+['model-end'] for x in labels]

model = pg.HiddenMarkovModel.from_samples(
                                          distribution=pg.IndependentComponentsDistribution([pg.PoissonDistribution, pg.NormalDistribution, pg.NormalDistribution]),
                                          n_components=len(hidden_states_types), # which is 5
                                          X = np.array(sequences), #the aforementioned sequence
                                          labels=np.array(labels), # a sequence of 10 x 24 x 1 which reflects the label for each hour
                                          state_names=hidden_states_types, 
                                          algorithm='labeled',
                                          name="model",
                                          verbose=True)
```

Running this code snippet returns the following error:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
~\AppData\Local\Temp/ipykernel_461836/4035827265.py in <module>
     17                                           algorithm='labeled',
     18                                           name="model",
---> 19                                           verbose=True)

~\Anaconda3\envs\env_pom\lib\site-packages\pomegranate\hmm.pyx in pomegranate.hmm.HiddenMarkovModel.from_samples()

~\Anaconda3\envs\env_pom\lib\site-packages\pomegranate\distributions\IndependentComponentsDistribution.pyx in pomegranate.distributions.IndependentComponentsDistribution.IndependentComponentsDistribution.from_samples()

ValueError: must pass in a list of distribution callables

```

If I change the distribution to `MultivariateGaussianDistribution`, it would work properly. What I am looking for though is to feed the distribution of each feature to the model and that is why I chose `IndependentComponentsDistribution`.

Any feedback/clue on what is going wrong here?

**Comment by jmschrei** [comment]
Howdy. I think the issue is that you're initializing the IndependentComponentsDistribution incorrectly. You need to pass in a list of initiated distributions, i.e. with parameters. However, if you're using the `from_samples` method, these initial parameters will be overwritten (and not affect the values they are overwritten with) in the first step of the training procedure so it doesn't matter what you set them to. Sorry about the inconvenience.

**Comment by mmrahimi** [comment]
Many thanks for your help, Jacob.
I also had tried this before. In that case, I'd get this error:

```
~\Anaconda3\envs\env_wpfour\lib\site-packages\pomegranate\hmm.pyx in pomegranate.hmm.HiddenMarkovModel.from_samples()

~\Anaconda3\envs\env_wpfour\lib\site-packages\pomegranate\distributions\IndependentComponentsDistribution.pyx in pomegranate.distributions.IndependentComponentsDistribution.IndependentComponentsDistribution.from_samples()

ValueError: must pass in a list of distribution callables
```

My code snippet is something like this:

```
dist = pg.IndependentComponentsDistribution([pg.PoissonDistribution(1), pg.NormalDistribution(0,1), pg.NormalDistribution(0,1)])

model = pg.HiddenMarkovModel.from_samples(distribution= dist,
                                          n_components=len(hidden_states_types), 
                                          X = np.array(sequences), 
                                          labels=np.array(labels),
                                          state_names=hidden_states_types,
                                          algorithm='labeled',
                                          name="model",
                                          verbose=True)

```

Any feedback is highly appreciated.

**Comment by jmschrei** [comment]
Ah, sorry, I forgot. In that case, I think the right thing to do is pass the distributions in directly as so:

```python
dist = pg.PoissonDistribution, pg.NormalDistribution, pg.NormalDistribution

model = pg.HiddenMarkovModel.from_samples(distribution= dist,
                                          n_components=len(hidden_states_types), 
                                          X = np.array(sequences), 
                                          labels=np.array(labels),
                                          state_names=hidden_states_types,
                                          algorithm='labeled',
                                          name="model",
                                          verbose=True)
```

Let me know if that works.

**Comment by mmrahimi** [comment]
No, I'm afraid it still got problems:

```
~\Anaconda3\envs\env_wpfour\lib\site-packages\pomegranate\hmm.pyx in pomegranate.hmm.HiddenMarkovModel.from_samples()

AttributeError: 'tuple' object has no attribute 'from_samples'
```

**Comment by jmschrei** [comment]
Sorry, dist needed to be a list. 

```Python
dist = [pg.PoissonDistribution, pg.NormalDistribution, pg.NormalDistribution]
```

**Comment by mmrahimi** [comment]
Sorry Jacob, this time it returns this:

`AttributeError: 'list' object has no attribute 'from_samples'`

I guess there should be `IndependentComponentsDistribution` or something similar which has the method `from_samples`. What do you think?

**Comment by jmschrei** [comment]
Can you post the full script and error message please? And can you try without labels? https://github.com/jmschrei/pomegranate/blob/master/pomegranate/hmm.pyx#L3669 should be the line that turns a list of distributions into an independent components distribution.

**Comment by mmrahimi** [comment]
Sure, I will try to mock up the data and share it here tonight. Here is the full error message:

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
~\AppData\Local\Temp/ipykernel_58008/3318621695.py in <module>
     20                                           algorithm='labeled',
     21                                           name="model",
---> 22                                           verbose=True)

~\Anaconda3\envs\env_wpfour\lib\site-packages\pomegranate\hmm.pyx in pomegranate.hmm.HiddenMarkovModel.from_samples()

AttributeError: 'list' object has no attribute 'from_samples'
```

BTW, I tried it without the labels, and it works like a charm. Something like this:

```
model = pg.HiddenMarkovModel.from_samples(distribution= dist,
                                          n_components=len(hidden_states_types), 
                                          X = combined_sequences_train, 
                                          state_names=hidden_states_types,
                                          name="model",
                                          verbose=True)

```
I suppose the issue happens when labels are being introduced.

**Comment by mmrahimi** [comment]
Meanwhile, I went through the `from_samples` method on `hmm.pyx` and inspired by your own work, I added the following snippet before [Line 3610](https://github.com/jmschrei/pomegranate/blob/master/pomegranate/hmm.pyx#L3610). It seems to be working now:

```
elif (isinstance(distribution, list)):
                distributions = []
                for label in label_set:
                    idx = labels_concat == label
                    distributions.append(
                        IndependentComponentsDistribution.from_samples(
                            X_concat[idx], 
                            distributions=distribution
                            )
                    )
```

What do you think?

**Comment by jmschrei** [comment]
I think that's correct. I'll take a stab at fixing the code and pushing a PR soon.

**Comment by jmschrei** [comment]
Okay. I think that I fixed it. I ended up separating the code for labeling the data initially (either pulling out the user-specified labeled data or using k-means to cluster an unlabeled data set) and for using the labeled data and a user specification to initialize a set of distributions. Hopefully, that clears up broader issues with the code-base.

**Comment by jmschrei** [comment]
Please re-open if not fixed in 0.14.6.
